presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,830 @@
+            allNulls = true;
+        }
+        else {
+            nestedLengths = ensureCapacity(nestedLengths, nonNullPositionCount);
+            Arrays.fill(nestedLengths, 0);
+
+            for (int keyIndex = 0; keyIndex < keyCount; keyIndex++) {
+                BooleanInputStream inMapStream = inMapStreams.get(keyIndex);
+
+                nestedPositions[keyIndex] = ensureCapacity(nestedPositions[keyIndex], nonNullPositionCount);
+                inMap[keyIndex] = ensureCapacity(inMap[keyIndex], nonNullPositionCount);
+
+                int nestedStreamPosition = 0;
+
+                int nestedSkipped = 0;
+                int nestedPositionCount = 0;
+                for (int i = 0; i < nonNullPositionCount; i++) {
+                    int position = nonNullPositions[i];
+                    if (position > nestedStreamPosition) {
+                        nestedSkipped += inMapStream.countBitsSet(position - nestedStreamPosition);
+                        nestedStreamPosition = position;
+                    }
+
+                    nestedStreamPosition++;
+                    if (inMapStream.nextBit()) {
+                        nestedPositions[keyIndex][nestedPositionCount] = nestedPositionCount + nestedSkipped;
+                        nestedPositionCount++;
+
+                        nestedLengths[i]++;
+                        inMap[keyIndex][i] = true;
+                    }
+                    else {
+                        inMap[keyIndex][i] = false;
+                    }
+                }
+
+                if (nonNullSkipped > nestedStreamPosition - nonNullPositionCount) {
+                    inMapStream.skip(nonNullSkipped - (nestedStreamPosition - nonNullPositionCount));
+                }
+
+                nestedPositionCounts[keyIndex] = nestedPositionCount;
+
+                if (nestedPositionCount > 0) {
+                    int readCount = valueStreamReaders.get(keyIndex).read(nestedReadOffsets[keyIndex], nestedPositions[keyIndex], nestedPositionCount);
+                    verify(readCount == nestedPositionCount);
+                }
+                nestedReadOffsets[keyIndex] += nestedSkipped + nestedPositionCount;
+            }
+
+            if (nullsAllowed) {
+                outputPositionCount = positionCount;
+            }
+            else {
+                System.arraycopy(nonNullPositions, 0, outputPositions, 0, nonNullPositionCount);
+                outputPositionCount = nonNullPositionCount;
+            }
+        }
+
+        populateNestedOutputPositions();
+    }
+
+    private void openRowGroup()
+            throws IOException
+    {
+        presentStream = presentStreamSource.openStream();
+
+        for (int i = 0; i < keyCount; i++) {
+            BooleanInputStream inMapStream = requireNonNull(inMapStreamSources.get(i).openStream(), "missing inMapStream at position " + i);
+            inMapStreams.add(inMapStream);
```

Comment:
Here we are trusting that the length of inMapStreams == i right when we add? So that we don't need to specify an index to add at?

Reply:
Yes, we rely on `inMapStreams.clear();` in `startRowGroup`